pkg/services/grafana-apiserver/storage/entity/utils.go

```diff
@@ -0,0 +1,263 @@
+package entity
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apiserver/pkg/endpoints/request"
+
+	"github.com/grafana/grafana/pkg/infra/grn"
+	"github.com/grafana/grafana/pkg/kinds"
+	entityStore "github.com/grafana/grafana/pkg/services/store/entity"
+)
+
+type Key struct {
+	Group       string
+	Resource    string
+	Namespace   string
+	Name        string
+	Subresource string
+}
+
+func ParseKey(key string) (*Key, error) {
+	// /<group>/<resource>/<namespace>/<name>(/<subresource>)
+	parts := strings.SplitN(key, "/", 6)
+	if len(parts) != 5 && len(parts) != 6 {
+		return nil, fmt.Errorf("invalid key (expecting 4 or 5 parts) " + key)
+	}
+
+	if parts[0] != "" {
+		return nil, fmt.Errorf("invalid key (expecting leading slash) " + key)
+	}
+
+	k := &Key{
+		Group:     parts[1],
+		Resource:  parts[2],
+		Namespace: parts[3],
+		Name:      parts[4],
+	}
+
+	if len(parts) == 6 {
+		k.Subresource = parts[5]
+	}
+
+	return k, nil
+}
+
+func (k *Key) String() string {
+	if len(k.Subresource) > 0 {
+		return fmt.Sprintf("/%s/%s/%s/%s/%s", k.Group, k.Resource, k.Namespace, k.Name, k.Subresource)
+	}
+	return fmt.Sprintf("/%s/%s/%s/%s", k.Group, k.Resource, k.Namespace, k.Name)
+}
+
+func (k *Key) IsEqual(other *Key) bool {
+	return k.Group == other.Group &&
+		k.Resource == other.Resource &&
+		k.Namespace == other.Namespace &&
+		k.Name == other.Name &&
+		k.Subresource == other.Subresource
+}
+
+func (k *Key) TenantID() (int64, error) {
+	if k.Namespace == "default" {
```

Comment:
it seems like we'd need to require a stack ID in cloud deploys. is `default` only valid for on-prem?

Reply:
yeah, we need to reconcile this with the work in grafana-apiserver/endpoints/request

Reply:
does storage even need to know the tenantId?  if it works with namespaces directly we would not need to parse it

Reply:
we use it to set the tenant id in the backend, and will use it for authn/z enforcement to prevent cross-tenant requests.